pkg/controller/keyvault/gateway_secret_provider_class.go

```diff
@@ -0,0 +1,231 @@
+				logger.Error(err, fullErr.Error())
+				g.events.Event(gwObj, corev1.EventTypeWarning, "FailedUpdateOrCreateSPC", fullErr.Error())
+				return ctrl.Result{}, fullErr
+			}
+
+			logger.Info(fmt.Sprintf("preemptively attaching secret reference for listener %s", listener.Name))
+			newCertRef := gatewayv1.SecretObjectReference{
+				Namespace: to.Ptr(gatewayv1.Namespace(req.Namespace)),
+				Group:     to.Ptr(gatewayv1.Group(corev1.GroupName)),
+				Kind:      to.Ptr(gatewayv1.Kind("Secret")),
+				Name:      gatewayv1.ObjectName(generateGwListenerCertName(gwObj.Name, listener.Name)),
+			}
+			gwObj.Spec.Listeners[index].TLS.CertificateRefs = []gatewayv1.SecretObjectReference{newCertRef}
+			continue
+		}
+		// we should delete the SPC if it exists
+		logger.Info(fmt.Sprintf("attempting to remove unused SPC %s", spc.Name))
+
+		deletionSpc := &secv1.SecretProviderClass{}
+		if err = g.client.Get(ctx, client.ObjectKeyFromObject(spc), deletionSpc); err != nil {
+			if client.IgnoreNotFound(err) != nil {
+				logger.Error(err, fmt.Sprintf("failed to fetch SPC for deletion %s", spc.Name))
+				return ctrl.Result{}, fmt.Errorf("fetching SPC for deletion: %w", err)
+			}
+			continue
+		}
+
+		if manifests.HasTopLevelLabels(deletionSpc.Labels) {
+			// return if we fail to delete, but otherwise, keep going
+			if err = g.client.Delete(ctx, deletionSpc); err != nil {
+				if client.IgnoreNotFound(err) != nil {
+					logger.Error(err, fmt.Sprintf("failed to delete SPC %s", spc.Name))
+					return ctrl.Result{}, fmt.Errorf("deleting SPC: %w", err)
+				}
+				continue
+			}
+		}
+
+	}
+
+	logger.Info("reconciling Gateway resource with new secret refs for each TLS-enabled listener")
+	if err = util.Upsert(ctx, g.client, gwObj); err != nil {
+		fullErr := fmt.Errorf("failed to reconcile Gateway resource %s: %w", req.Name, err)
+		logger.Error(err, fullErr.Error())
+		g.events.Event(gwObj, corev1.EventTypeWarning, "FailedUpdateOrCreateGateway", fullErr.Error())
+		return ctrl.Result{}, fullErr
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func generateGwListenerCertName(gw string, listener gatewayv1.SectionName) string {
+	certName := fmt.Sprintf("kv-gw-cert-%s-%s", gw, string(listener))
+	if len(certName) > 253 {
+		certName = certName[:253]
+	}
+
+	return certName
+}
+
+func listenerIsKvEnabled(listener gatewayv1.Listener) bool {
+	return listener.TLS != nil && listener.TLS.Options != nil && listener.TLS.Options[tlsCertKvUriAnnotation] != ""
+}
+
+func retrieveClientIdForListener(ctx context.Context, k8sclient client.Client, namespace string, options map[gatewayv1.AnnotationKey]gatewayv1.AnnotationValue) (string, error) {
+	certUri := string(options[certUriTLSOption])
+	saName := string(options[serviceAccountTLSOption])
+
+	// validate user input
+	if certUri != "" && saName == "" {
+		return "", newUserError(errors.New("user specified cert URI but no ServiceAccount in a listener"), "detected Keyvault Cert URI, but no ServiceAccount was provided")
```

Comment:
detected seems like the wrong word here. Tell the user exactly what they need to do to resolve the problem.

"KeyVault Cert URI provided but the required ServiceAccount option was not. Both are required for the KeyVault TLS experience Provide the ServiceAccount by doing x"